File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque
N, K = map(int, input().split(" "))

visited = [0 for _ in range(100_001)]

def BFS(n):
    q = deque()
    q.append(n)
    while q:
        mq = q.popleft()
        move = [mq-1, mq+1, mq*2]
        if mq == K:
            return visited[mq]
        for i in range(3):
            m = move[i]
            if 0 <= m <= 100_000 and not visited[m]:
                visited[m] = visited[mq] + 1
                q.append(m)
print(BFS(N))
# print(visited[:20])

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;

public class Main {
	
	static int max = 100001;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] NM = br.readLine().split(" ");
		int N = Integer.parseInt(NM[0]);
		int K = Integer.parseInt(NM[1]);
		
		bfs(N, K);
	}
	
	private static void bfs(int start, int K) {
		Queue<Integer> queue = new ArrayDeque<>();
		boolean[] visited = new boolean[max];
		int[] count = new int[max];
		queue.add(start);
		visited[start] = true;
		while(!queue.isEmpty()) {
			int now = queue.poll();
			if(now == K) {
				break;
			}
			int[] move = {now + 1, now - 1, now * 2};
			for(int i : move) {
				if(0<= i && i < max) {
					if(!visited[i]) {
						visited[i] = true;
						count[i] = count[now] + 1;
						queue.add(i);
					}					
				}
			}
		}
		System.out.println(count[K]);
	}

}